Async Events

oersted opened the pull request:
## What is the motivation?

It allows reacting to database changes without blocking write requests.

It is possible to replicate such behavior with a client subscribing to live queries of course, but having it embedded in SurrealDB is less complex and more efficient in many ways. SurrealQL is plenty powerful and flexible for common use-cases, the collection of built-in functions is rich, and the ability to embed JavaScript can fill the gaps.

At present, the need to enrich databases or take action on data with chains of external API requests is growing, both for calling LLMs or integrating with various services. Such API requests may be long-running. Such reactive events will also often write back to the DB, which may trigger other event chains. Having all these events be part of the same transaction and blocking the original request is quite limiting, in some use-cases it may be fine or desirable to have these reactions be eventually-consistent.

More concretely, the issue #2923 stated a case of replacing a serverless Firestore + Functions system, where reactive chains may lead to 10-50 writes.

> Current use case is creating serverless functions that run in bun and run in gke pods, for now its fine as they are all 1-2ms(in the test project), but this will easily grow when some chains of changes amount into 10-50 writes that could all have their own events. Migrating away from Firestore/gcp serverless functions which fire off the write events all at once into a queue that's fire and forget as far as the database is concerned.

Personally, I'd like to use SurrealDB for reactive data pipelines, where new data gets enriched in complex ways from various external sources. I'd also like to adopt SurrealDB into a project where we are building an experimental game/simulation engine with a persistent world-state and real-time ECS-style decoupled systems that independently react to world events to implement logic, often involving AI.

This gets somewhat into the territory of [SpacetimeDB](https://github.com/clockworklabs/SpacetimeDB) or [Materialize](https://github.com/MaterializeInc/materialize), two other Rust-based innovative DBs, and the database-as-a-backend pattern is becoming increasingly common. I prefer SurrealDB for quite a few reasons and this feature will help close the gap and be more competitive towards those alternatives.


## What does this change do?

It adds the ability to `DEFINE EVENT [...] CONCURRENTLY` (~`DEFINE EVENT ASYNC [...]`~).

Currently events run in the same transaction as the write request they are reacting to and block the original request until all events complete executing. Async events are run in a separate independent transaction without blocking the DB change they are reacting to.

**Implementation**:

- [x] Minimally extend the query parser and the event data model.
- [x] Execute events with `tokio::spawn`, without awaiting and dropping the `JoinHandle`, with `Context::new_concurrent`.

Currently this feature is not supported in WASM and fails gracefully in that case. `Context::new_concurrent` is not present in WASM and there are complexities with async in WASM I don't feel qualified to get into (see this [comment in the issue](https://github.com/surrealdb/surrealdb/issues/2923#issuecomment-1900093365)), I'd welcome ideas on this.

Additionally, in the current implementation async events are just fire-and-forget futures. This means they will be silently canceled when the process exits. Since they are wrapped in a transaction they will at least not lead to an inconsistent state, but there is no strong guarantee that they will complete. This [comment](https://github.com/surrealdb/surrealdb/issues/2923#issuecomment-3405995879) also raised the matter of keeping a log background events and their state in the DB. At least some better error logging seems important.

I'd welcome suggestions on this, I feel like adding such a metadata table for every DB is too big a change for me to tackle in my first PR, and it would be good to keep this feature lean to begin with. It can be documented as being experimental like a number of other features are right now.


## What is your testing strategy?

- [x] The full test suite passes (*)
- [ ] The full test suite passes for a WASM build
- [x] Extend the `DEFINE EVENT` parsing test for the `ASYNC` case
- [ ] Extend `language-tests` to validate that they actually run in a non-blocking manner

> (*) I had to exclude a few of the integration test suites (`graphql`, `http`, `cli`) because I couldn't get them to pass on a clean clone. I reviewed the available documentation and run the tests with the recommended features, but there are still issues. At least I checked that my changes don't make any new ones fail. I'd appreciate some help to run the full test suite properly.

## Is this related to any issues?

Closes #2923

## Does this change need documentation?

Yes mainly here: https://surrealdb.com/docs/surrealql/statements/define/event

> I haven't written a PR at the [docs repo](https://github.com/surrealdb/docs.surrealdb.com) yet

## Does this change make any alterations to environment variables or CLI commands?

- [x] No changes made to env vars

## Have you read the Contributing Guidelines?

- [x] I have read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md)

oersted committed:
You can now "DEFINE EVENT ASYNC ...", it does nothing for now.

Added is_async to EventDefinition and DefineEventStatement in expr and sql, adapted the parser accordingly, and extended the relevant tests to validate it.

oersted committed:
Implement execution of async events

The test suite passed for the non-async case, I will add async event tests and debug next.

It tokio::spaws the event execution future and drops the join handle, such that it keeps running as a background task. It is wrapped in an independent transaction (Context) and executed with a fresh stack (Stk).

Async events are not supported in WASM for now. Added a syntax error when doing DEFINE EVENT ASYNC in WASM. If an async event is executed in WASM regardless, it will be run synchronously like non-async events with a tracing::warn.

oersted commented on crates/core/src/syn/lexer/keywords.rs:
```diff
@@ -71,6 +71,7 @@ pub(crate) static KEYWORDS: phf::Map<UniCase<&'static str>, TokenKind> = phf_map
 	UniCase::ascii("ASC") => TokenKind::Keyword(Keyword::Ascending),
 	UniCase::ascii("ASCII") => TokenKind::Keyword(Keyword::Ascii),
 	UniCase::ascii("ASSERT") => TokenKind::Keyword(Keyword::Assert),
+	UniCase::ascii("ASYNC") => TokenKind::Keyword(Keyword::Async),
```
I was thinking the same actually, I just copied the suggested syntax from #2923, but I felt like the ASYNC term refers more to the implementation rather than the behavior from the users perspective. At least it's concise and easy to read.

The intent is for it not to block for the original request transaction (or for the original request not to wait for the event), so it could also be something like NONBLOCKING, NOWAIT or BACKGROUND.

There was already some confusion as to what "concurrent" meant in this context in the original issue: https://github.com/surrealdb/surrealdb/issues/2923#issuecomment-1910203663

But it does make sense that it matches the indexing case, since we use the same `Context::new_concurrent` for this, although I'm not sure if the actual execution behavior is the same. By the way, I would appreciate some feedback on that, the method seemed like a good fit but not sure if there's some subtle issue applying it here.

Since it may be a lasting user-facing addition to the syntax, which maintainer should take the decision on this?

LucyEgan commented on crates/core/src/doc/event.rs:
```diff
@@ -59,24 +68,51 @@ impl Document {
+			{
+				warn!(
+					"ASYNC events are not supported in WASM, they will run synchronously like non-async events"
+				);
```
Wouldn't this warn at the point of trigging any event in wasm and not just for async.
And from my understanding `DEFINE ASYNC` errors on wasm so they are not able to define the event to start with? (crates/core/src/syn/parser/stmt/define.rs)

oersted replied:
You are right, I missed the `if ev.is_async`, I added it in a778dc3.

It's true that `DEFINE EVENT ASYNC` throws a `SyntaxError` in WASM, but I wasn't sure if there may be a case where such an event might slip through (through an import or something like that). I didn't want it to completely silently ignore the ASYNC keyword, so I thought a simple warning would be appropriate as a fallback.

oersted commented:
@kearfy Of course, I understand, that sounds good, I'll continue polishing it so that it's ready by then.

Is there anywhere I can keep track with the release roadmap and/or timeline?

> this feature has been something on our radar for a longer time

It would be great to have more context on that as well, so that other use-cases are taken into account.

oersted committed:
Only warn when running async events in WASM if event is async

Also polished the syntax error when using ASYNC in WASM, it doesn't make sense to say it will run synchronously if the syntax is not allowed.

oersted committed:
Renamed ASYNC -> CONCURRENTLY and made consistent with DEFINE INDEX

The CONCURRENTLY keyword already exists for a similar purpose, best to reuse it for simplifity.

tobiemh commented:
Thanks for opening this PR @oersted 🚀 . This functionality should already be available in SurrealDB 3.0 so as a result I'm going to close the issue 😃 🎉
